



go1.13为 *errors* 和 *fmt* 标准库包引入了新特性，以简化处理包含其他错误的错误。其中最重要的是: 包含另一个错误的 *error* 可以实现返回底层错误的 *Unwrap* 方法。如果 *e1.Unwrap()* 返回 *e2*，那么我们说 *e1* 包装 *e2*，您可以展开 *e1* 以获得 *e2*。

按照此约定，我们可以为上面的 *QueryError* 类型指定一个 *Unwrap* 方法，该方法返回其包含的错误:



**go1.13 *errors* 包包含两个用于检查错误的新函数：*Is* 和 *As*。**





# **Wrapping errors with %w**



如前所述，使用 *fmt.Errorf* 向错误添加附加信息。



在 Go 1.13中 *fmt.Errorf* 支持新的 *%w* 谓词。



用 *%w* 包装错误可用于 *errors.Is* 以及 *errors.As*:



![image-20220822171536933](/Users/kestrel/developer/nrookie.github.io/collections/go/xunlianying/errors/image-20220822171536933.png)



作业



我们在数据库操作的时候， 比如dao层中当遇到一个sql.ErrNoRows 的时候， 是否应该Wrap这个Error， 抛给上层。 为什么， 应该怎么做请写出代码？

